---
title: "lab3_assembly_annotation"
output: html_document
date: "2025-09-18"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

```{r}
library("IRanges")
library("GenomicRanges")
library("Rsamtools")
library("GenomicAlignments")
library("Gviz")
```

Q1. **JUST MANUALLY GO TO END
Assembly statistics:

	Total length:	10245769
	Fragments:	432
	Fragments N50:	28892
	Largest frg:	85173
	Scaffolds:	0
	Mean coverage:	6

Based on the summary alignment statistics (found at the end of the flye_output/flye.log), we must look at a few things to consider if the coverage of the assembly is adequate. Given the total length of the assembly is over 10 million and the N50 is 

Q2.
i) BUSCO SCORES
C:37.9%[S:37.8%,D:0.1%],F:0.6%,M:61.5%,n:3282,E:5.8%
C - lvl of completeness (S = unique, D = duplicated)
The gene is uniquely found within 37.8% of the comparison of assemblies and has minimal duplication
F - the lvl of fragmentation = 0.6% -> limited fragmentation means more complete assemblies
M - lvl of how much the assembly is missing = 61.5% -> huge proportion of gene is unknown 
Given the coverage, although the assembly of the gene that is known is unique for the most part, a large portion of the assembly is unknown indicating coverage is limited and is not the best

ii/iii) ASSEMBLY STATISTICS
Assembly Statistics:
	432	Number of scaffolds
	432	Number of contigs
	10245769	Total length
	0.000%	Percent gaps
	28 kbp	Scaffold N50
	28 kbp	Contigs N50
The N50 score is 28kbp and compared to the number of total contigs run, 

	
Q3. (BELOW IS THE PORTION OF THE REPORT COPY AND PASTED)

Statistics without reference	assembly
# contigs	432
# contigs (>= 0 bp)	432
# contigs (>= 1000 bp)	432
# contigs (>= 5000 bp)	428
# contigs (>= 10000 bp)	374
# contigs (>= 25000 bp)	165
# contigs (>= 50000 bp)	26
Largest contig	85173
Total length	10245769
Total length (>= 0 bp)	10245769
Total length (>= 1000 bp)	10245769
Total length (>= 5000 bp)	10227184
Total length (>= 10000 bp)	9796199
Total length (>= 25000 bp)	6306256
Total length (>= 50000 bp)	1605042
N50	28892
N90	12557
auN	32584
L50	122
L90	324
GC (%)	41.87
Mismatches	
# N's per 100 kbp	0
# N's	0

The genome fraction 

Q4.
```{r}
library(Biostrings)

#load the full genome FASTA
genome <- readDNAStringSet("/Users/aaronsantos/GCF_000146045.2_R64_genomic.fna")

#extract chrom
chr1 <- genome[grep("chromosome I", names(genome), ignore.case = TRUE)]

#save
writeXStringSet(chr1, "R64_chr1.fna")

```

VISUALIZING ALIGNMENT TO CHROM 1 w/GRanges
Step 1. Import bam file + convert into Granges obj
```{r}

BiocManager::install("Rsamtools")

# load package
library(Rsamtools)

```

```{r}
BiocManager::install("GenomicAlignments")

library(GenomicAlignments)
```

```{r}
#BAM file path 
bam_file <- "/Users/aaronsantos/alignment.sorted.bam"

#open connection
bam <- BamFile(bam_file)

#read all alignments
aligns <- readGAlignments(bam)

gr_aligns <- granges(aligns)
```

Step 2. Create "tracks" showing alignment
```{r}
BiocManager::install("Gviz")

library(Gviz)
```

```{r}
#disable uscs-file name syntax
options(ucscChromosomeNames=FALSE)

align_track <- AnnotationTrack(gr_aligns, name = "Contigs", genome = "sacCer3", chromosome = "chr1")

axis_track <- GenomeAxisTrack()
```

Step 3. Visualize the alignment
```{r}
library(Gviz)

#TROUBLESHOOTING -> referred to ChatGPT as nothing was showing up
# make sure chromosome matches your GRanges
#chrom <- as.character(seqnames(gr_aligns)[1])

align_track <- AnnotationTrack(
    gr_aligns,
    name = "Contigs",
    genome = "sacCer3",
)

axis_track <- GenomeAxisTrack()

plotTracks(
    list(axis_track, align_track),
    from = min(start(gr_aligns)),
    to   = max(end(gr_aligns)),
    main = "Contig Alignments to Chromosome 1"
)

```

Q5.
i) Number of contigs
```{r}
length(gr_aligns)
```
ii) Plot
```{r}
plotTracks(
    list(axis_track, align_track),
    from = min(start(gr_aligns)),
    to   = max(end(gr_aligns)),
    main = "Contig Alignments to Chromosome 1"
)
```

Q6.
```{r}
library(Biostrings)

#file path
assembly_file <- "/Users/aaronsantos/assembly.fasta"

contigs <- readDNAStringSet(assembly_file)

#find lengths of each contig
contig_lengths <- width(contigs)

#get longest contig
longest_index <- which.max(contig_lengths)

#extract the longest contig
longest_contig <- contigs[longest_index]

writeXStringSet(longest_contig, "longest_contig.fasta")
```

Q7.
```{bash, eval=FALSE}

#convert sam to bam file
samtools view -bS /Users/aaronsantos/cds_aln.sam > cds_aln.bam

#sort aligned contigs
samtools sort cds_aln.bam -o cds_aln.sorted.bam

#index sorted bam -> creates .bai file
samtools index cds_aln.sorted.bam
```

Q8.
```{r}
#import bam file
#file path
bam_file2 <- "/Users/aaronsantos/cds_aln.sorted.bam"

#open cnnxn to bam file
bam2 <- BamFile(bam_file2)

#read all alignments
aligns2 <- readGAlignments(bam2)

#convert to GRanges obj
gr_aligns2 <- granges(aligns2)

#create tracks
align_track2 <- AnnotationTrack(
    gr_aligns2,
    name = "Contigs",
    genome = "sacCer3"
)

axis_track2 <- GenomeAxisTrack()

plotTracks(
    list(axis_track2, align_track2),
    from = min(start(gr_aligns2)),
    to   = max(end(gr_aligns2)),
    main = "Contig Alignments"
)
```

